File: legacy/nodes/core/llm/tools_builder_node.py
from typing import Any

from core.types_registry import (
    LLMToolSpec,
    NodeCategory,
    NodeInputs,
    get_type,
    serialize_for_api,
    validate_llm_tool_spec,
)
from nodes.base.base_node import Base


class ToolsBuilder(Base):
    """
    Takes a list of max 5 tool specs and outputs them as LLMToolSpecList.

    Input:
    - tools_list: List[LLMToolSpec] - less than 5 tool specifications

    Output:
    - tools: LLMToolSpecList
    """

    inputs = {
        "tools_list": get_type("LLMToolSpecList"),
    }

    optional_inputs = ["tools_list"]

    outputs = {
        "tools": get_type("LLMToolSpecList"),
    }

    CATEGORY = NodeCategory.LLM

    async def _execute_impl(self, inputs: NodeInputs) -> dict[str, Any]:
        tools_input = inputs.get("tools_list", [])

        if len(tools_input) > 5:
            raise ValueError(f"tools_list must be less than 5 tools, got {len(tools_input)}")

        # Validate and convert each tool spec to LLMToolSpec Pydantic model
        validated_tools: list[LLMToolSpec] = []
        for i, tool_spec in enumerate(tools_input):
            validated = validate_llm_tool_spec(tool_spec)
            if validated:
                validated_tools.append(validated)
            else:
                raise ValueError(f"Tool at index {i} is invalid: expected LLMToolSpec")

        return {"tools": serialize_for_api(validated_tools)}

File: nodes/llm/tools-builder-node.ts
// src/nodes/core/llm/tools-builder-node.ts
// Translated from: nodes/core/llm/tools_builder_node.py

import { Base, NodeCategory, getType, LLMToolSpec, validateLLMToolSpec, serializeForApi } from '@fig-node/core';
import type { NodeUIConfig } from '@fig-node/core';

/**
 * Takes a list of max 5 tool specs and outputs them as LLMToolSpecList.
 *
 * Input:
 * - tools_list: List[LLMToolSpec] - less than 5 tool specifications
 *
 * Output:
 * - tools: LLMToolSpecList
 */
export class ToolsBuilder extends Base {
  static inputs = {
    tools_list: getType('LLMToolSpecList'),
  };

  static optional_inputs = ['tools_list'];

  static outputs = {
    tools: getType('LLMToolSpecList'),
  };

  static CATEGORY = NodeCategory.LLM;

  static uiConfig: NodeUIConfig = {
    size: [200, 80],
    displayResults: false,
    resizable: false,
  };

  protected async executeImpl(
    inputs: Record<string, unknown>
  ): Promise<Record<string, unknown>> {
    const toolsInput = (inputs.tools_list as LLMToolSpec[]) || [];

    if (toolsInput.length > 5) {
      throw new Error(`tools_list must be less than 5 tools, got ${toolsInput.length}`);
    }

    // Validate and convert each tool spec
    const validatedTools: LLMToolSpec[] = [];

    for (let i = 0; i < toolsInput.length; i++) {
      const toolSpec = toolsInput[i];
      const validated = validateLLMToolSpec(toolSpec);

      if (validated) {
        validatedTools.push(validated);
      } else {
        throw new Error(`Tool at index ${i} is invalid: expected LLMToolSpec`);
      }
    }

    return { tools: serializeForApi(validatedTools) };
  }
}
